File: project_java/src/main/java/hu/bme/aut/onlab/dxhjoh/couchsurfing_java_backend/mapper/UserMapper.java
package hu.bme.aut.onlab.dxhjoh.couchsurfing_java_backend.mapper;

import hu.bme.aut.onlab.dxhjoh.couchsurfing_java_backend.dto_deprecated.UserDto;
import hu.bme.aut.onlab.dxhjoh.couchsurfing_java_backend.model.User;
import hu.bme.aut.onlab.dxhjoh.couchsurfing_java_backend.request.UserRequest;
import hu.bme.aut.onlab.dxhjoh.couchsurfing_java_backend.response.UserResponse;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.MappingTarget;
import org.mapstruct.ReportingPolicy;

import java.util.List;

@Mapper(componentModel = "spring", unmappedTargetPolicy = ReportingPolicy.IGNORE, uses = { BookingMapper.class, RoomMapper.class })
public interface UserMapper {

    UserResponse toResponse(User entity);

    List<UserResponse> toResponseList(List<User> entity);

    User toEntity(UserRequest request);

    User update(@MappingTarget User entity, UserRequest request);
}

File: project/src/main/kotlin/hu/bme/aut/onlab/couchsurfing_backend/mapper/UserMapper.kt
package hu.bme.aut.onlab.couchsurfing_backend.mapper

import hu.bme.aut.onlab.couchsurfing_backend.model.User
import org.mapstruct.Mapper
import org.mapstruct.Mapping
import org.mapstruct.ReportingPolicy
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.crypto.password.PasswordEncoder

@Mapper(componentModel = "spring", unmappedTargetPolicy = ReportingPolicy.IGNORE)
abstract class UserMapper(@Autowired private var passwordEncoder: PasswordEncoder) {

    @Mapping(expression = "java(getRooms(entity))", target = "rooms")
    @Mapping(target = "password", ignore = true) // hibát dob: Repeatable annotations with non-SOURCE retention are not yet supported
    // elvileg ez is nyelvi sajátosság, hogy JDK 1.6-ra targetál a Kotlin, és ott nem lehet ilyen repeatable annotációkat használni
    // https://stackoverflow.com/questions/44543978/kotlin-repeatable-annotations-dont-work-on-jdk-8
    abstract fun toDto(entity: User)
}